orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc

```diff
@@ -194,7 +193,7 @@ Status SoftmaxCrossEntropyLoss<T1, T2>::Compute(OpKernelContext* context) const
     if (reduction_ != ReductionType::NONE) {
       // Sum loss over n_d samples
       math::Sum<T1, CPUMathUtil>(n_d, loss_sample, loss_data, nullptr);
-      if (reduction_ == ReductionType::MEAN) {
+      if ((reduction_ == ReductionType::MEAN) && (unignored_samples != 0)) {
```

Comment:
In the case there is no valid sample in this batch, there is no way to compute the gradients. 
We should error out in the case.

Reply:
Also, do we have this problem for CUDA kernels?

Reply:
We don't have issue on CUDA because there we don't really divide total loss by zero as we have normalize_factor which will be zero but it won't be used at all as all values will be dropped (since target value will be same as ignore_index) and in the end we will do a reduce sum on all zero loss values which will result in a zero. The UT verifies this.  

Why is a gradient of zero not valid when there are no valid values?

Reply:
If all labels are dropped because of `ingore_index`, the loss and gradient should be 0. Let's consider a math equation of `ignore_index=0` and all labels are 0. The loss is actually
```
0 * loss(label[0], predict[0]) + 0 * loss(label[1], predict[1]) + ... + 0 * loss(label[L-1], predict[L-1])
```
where 
```
label[0] = label[1] = ... = label[L-1]
```
Because of the `0*` before individual loss term, their total loss is 0 and their gradient will be zero 0 (taking derivative against a constant 0 w.r.t. any tensor would be 0).

A batch with all label being `ignored` is still a valid data point. It can appear in user's data set, so we can't throw.

Reply:
Thanks, @wschin !!!